haystack/nodes/retriever/web.py

```diff
@@ -201,64 +106,53 @@ def retrieve(  # type: ignore[override]
-        :param cache_time: The time limit in seconds to check the cache. The default is 24 hours.
+        :param cache_time: The time limit in seconds for the documents in the cache. If objects are older than this time,
+        they will be deleted from the cache on the next retrieval.
         """
 
         # Initialize default parameters
         preprocessor = preprocessor or self.preprocessor
-        cache_document_store = cache_document_store or self.cache_document_store
         cache_index = cache_index or self.cache_index
+        top_k = top_k or self.top_k
         cache_headers = cache_headers or self.cache_headers
         cache_time = cache_time or self.cache_time
-        top_k = top_k or self.top_k
 
-        # Normalize query
-        query_norm = self._normalize_query(query)
+        search_results, _ = self.web_search.run(query=query)
+        result_docs = search_results["documents"]
 
-        # Check cache for query
-        extracted_docs = self._check_cache(
-            query_norm, cache_index=cache_index, cache_headers=cache_headers, cache_time=cache_time
-        )
+        if self.mode != "snippets":
+            # for raw_documents and preprocessed_documents modes, we need to retrieve the links from the search results
+            links: List[SearchResult] = self._prepare_links(result_docs)
 
-        # If query is not cached, fetch from web
-        if not extracted_docs:
-            extracted_docs = self._retrieve_from_web(query_norm, preprocessor)
-
-        # Save results to cache
-        if cache_document_store and extracted_docs:
-            cached = self._save_cache(query_norm, extracted_docs, cache_index=cache_index, cache_headers=cache_headers)
-            if not cached:
-                logger.warning(
-                    "Could not save retrieved Documents to the DocumentStore cache. "
-                    "Check your DocumentStore configuration."
-                )
-        return extracted_docs[:top_k]
+            links_found_in_cache, cached_docs = self._check_cache(links)
+            logger.debug("Found %d links in cache", len(links_found_in_cache))
 
-    def _retrieve_from_web(self, query_norm: str, preprocessor: Optional[PreProcessor]) -> List[Document]:
-        """
-        Retrieve Documents from the web based on the query.
+            links_to_fetch = [link for link in links if link not in links_found_in_cache]
+            logger.debug("Fetching %d links", len(links_to_fetch))
+            result_docs = self._scrape_links(links_to_fetch)
 
-        :param query_norm: The normalized query string.
-        :param preprocessor: The PreProcessor to be used to split documents into paragraphs.
-        :return: List of Document objects.
-        """
+            # Save result_docs to cache
+            self._save_to_cache(
+                result_docs, cache_index=cache_index, cache_headers=cache_headers, cache_time=cache_time
+            )
```

Comment:
Shouldn't we also update the timestamp of the already cached links? 🤔

Reply:
We don't need to, no need to complicate things IMHO. Let's just have them expire after ttl has expired...

Reply:
Good enough for me. 🤷